client/src/main/java/org/apache/nemo/client/beam/NemoPipelineResult.java

```diff
@@ -44,14 +59,36 @@ public State getState() {
 
   @Override
   public State cancel() throws IOException {
-    throw new UnsupportedOperationException("cancel() in frontend.beam.NemoPipelineResult");
+    try {
+      JobLauncher.shutdown();
+      this.jobDone.await();
+      return State.CANCELLED;
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
   }
 
   @Override
   public State waitUntilFinish(final Duration duration) {
-    throw new UnsupportedOperationException();
-    // TODO #208: NemoPipelineResult#waitUntilFinish hangs
-    // Previous code that hangs the job:
+    try {
+      if (duration.getMillis() < 1) {
+        this.jobDone.await();
+        return State.DONE;
+      } else {
+        final boolean finished = this.jobDone.await(duration.getMillis(), TimeUnit.MILLISECONDS);
+        if (finished) {
+          LOG.info("Job successfully finished before timeout of {}ms, while waiting until finish",
+            duration.getMillis());
+          return State.DONE;
+        } else {
+          LOG.warn("Job timed out before {}ms, while waiting until finish. Call 'cancel' to cancel the job.",
+            duration.getMillis());
+          return State.RUNNING;
+        }
+      }
+    } catch (final Exception e) {
+      throw new RuntimeException(e);
+    }
     // return (State) super.waitUntilJobFinish(duration.getMillis(), TimeUnit.MILLISECONDS);
```

Comment:
Why do we have this line commented out here?

Reply:
@johnyangk git blame tells me it's written by you?

Reply:
Okay so it comes along with the previous TODO, @wonook @johnyangk so perhaps we should keep it here?

Reply:
Yes, this wasn't part of the PR, so I think it could be handled on a separate issue.